File: medium/Path_Sum_II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if not root:
            return []
        ans = []
        def next_node(node, num_list, sum_num):
            num_list = num_list + [node.val]
            sum_num = sum_num + node.val
            if node.left is None and node.right is None:
                if sum_num == targetSum:
                    ans.append(num_list)
            else:
                if node.left is not None:
                    next_node(node.left, num_list, sum_num)
                if node.right is not None:
                    next_node(node.right, num_list, sum_num)
        next_node(root, [], 0)
        return ans

File: cpp/medium/Path_Sum_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> ans;
        if (root == nullptr) return ans;
        vector<int> comb;
        next_node(root, targetSum, 0, comb, ans);
        return ans;
    }
    void next_node(TreeNode* node, int targetSum, int current_sum, vector<int>& comb, vector<vector<int>>& ans) {
        comb.push_back(node->val);
        if (node->left == nullptr && node->right == nullptr) {
            if (current_sum + node->val == targetSum) {
                ans.push_back(comb);
            }
        }
        if (node->left != nullptr) {
            next_node(node->left, targetSum, current_sum + node->val, comb, ans);
            comb.pop_back();
        }
        if (node->right != nullptr) {
            next_node(node->right, targetSum, current_sum + node->val, comb, ans);
            comb.pop_back();
        }
    }
};
